src/hotspot/share/opto/loopnode.cpp

```diff
@@ -1616,6 +1616,88 @@ bool PhaseIdealLoop::convert_to_long_loop(Node* cmp, Node* phi, IdealLoopTree* l
 
 //------------------------------is_counted_loop--------------------------------
 bool PhaseIdealLoop::is_counted_loop(Node* x, IdealLoopTree*&loop, BasicType iv_bt) {
+  return do_is_counted_loop(x, loop, iv_bt) || is_counted_loop_with_speculative_long_limit(x, loop, iv_bt);
+}
+
+// When comparing an int incrementor with long limit, check if the limit is within int range. If so, treat it as an int
+// counted loop but deoptimize if the limit is out of int range.
+// This is common pattern with "for (int i =...; i < long_limit; ...)" where int "i" is implicitly promoted to long,
+// i.e., "(long) i < long_limit", and therefore making it not an int counted loop without this transformation.
+bool PhaseIdealLoop::is_counted_loop_with_speculative_long_limit(Node* x, IdealLoopTree*&loop, BasicType iv_bt) {
+  if (iv_bt != T_INT) {
+    return false;
+  }
+
+  Node* back_control = loop_exit_control(x, loop);
+  if (back_control == nullptr) {
+    return false;
+  }
+
+  Node* init_control = x->in(LoopNode::EntryControl);
+
+  // Make sure there is a parse predicate for us to insert the loop limit check.
+  const Predicates predicates(init_control);
+  const PredicateBlock* loop_limit_check_predicate_block = predicates.loop_limit_check_predicate_block();
+  if (!loop_limit_check_predicate_block->has_parse_predicate()) {
+    return false;
```

Comment:
I suggest to also add here a trace message, similar to what we already do in `is_counted_loop()`:
https://github.com/openjdk/jdk/blob/c330b90b9f43f80c322153585fa78704358f0224/src/hotspot/share/opto/loopnode.cpp#L2013-L2023